src/ROOTNTupleReader.cc

```diff
@@ -0,0 +1,245 @@
+      collBuffers.vectorMembers = new podio::VectorMembersInfo();
+      collBuffers.vectorMembers->reserve(vminfo->size());
+
+      for (const auto& [type, _] : (*vminfo)) {
+        const auto* vecClass = TClass::GetClass(("vector<" + type + ">").c_str());
+        collBuffers.vectorMembers->emplace_back(type, vecClass->New());
+      }
+    }
+
+    if (!isSubsetColl) {
+      dentry->CaptureValueUnsafe(m_collectionName[category][i], collBuffers.data);
+    }
+    if (auto* refCollections = collBuffers.references) {
+      for (size_t j = 0; j < refCollections->size(); ++j) {
+        // // The unique_ptrs are nullptrs at the beginning, we first initialize
+        // // them and then fill the values with the read data since
+        // refCollections->at(j) = std::make_unique<std::vector<podio::ObjectID>>();
+        // const auto brName = root_utils::refBranch(m_collectionName[category][i], j);
+        // std::cout << "brName = " << brName << " " << (refCollections->at(j) == nullptr) << std::endl;
+        // dentry->CaptureValueUnsafe(brName, (*refCollections)[j].get());
+
+        auto vec = new std::vector<podio::ObjectID>;
+        const auto brName = root_utils::refBranch(m_collectionName[category][i], j);
+        dentry->CaptureValueUnsafe(brName, vec);
+        tmp[{brName, j}] = vec;
+      }
+    }
+
+    if (auto* vecMembers = collBuffers.vectorMembers) {
+      for (size_t j = 0; j < vecMembers->size(); ++j) {
+        const auto typeName = "vector<" + vecMembers->at(j).first + ">";
+        const auto brName = root_utils::vecBranch(m_collectionName[category][i], j);
+        dentry->CaptureValueUnsafe(brName, vecMembers->at(j).second);
+      }
+    }
+
+    std::cout << "CaptureValueUnsafe done" << std::endl;
+    buffers.emplace(m_collectionName[category][i], std::move(collBuffers));
+  }
+
+  m_readers[category][0]->LoadEntry(entNum);
+
+  for (size_t i = 0; i < m_collectionId[category].size(); ++i) {
+    auto collBuffers = buffers[m_collectionName[category][i]];
+    if (auto* refCollections = collBuffers.references) {
+      for (size_t j = 0; j < refCollections->size(); ++j) {
+        const auto brName = root_utils::refBranch(m_collectionName[category][i], j);
+        refCollections->at(j) = std::unique_ptr<std::vector<podio::ObjectID>>(tmp[{brName, j}]);
+      }
+    }
+
+  }
+
+  auto parameters = readEventMetaData(category, entNum);
+  auto table = std::make_shared<CollectionIDTable>();
+
+  auto names = m_collectionName[category];
+  auto ids = m_collectionId[category];
+
+  // Sort the ids and collection names to recreate the collection ID table with
+  // the names following the order of the IDs
+  std::vector<std::pair<int, std::string>> v;
+  v.reserve(names.size());
+  for (size_t i = 0; i < names.size(); ++i) {
+    v.emplace_back(std::make_pair<int, std::string>(int(ids[i]), std::string(names[i])));
+  }
+  std::sort(v.begin(), v.end());
+  for (auto& [id, name] : v) {
+    table->add(name);
+  }
```

Comment:
This should be unnecessary, you can construct the `CollectionIDTable` directly from these vectors:
https://github.com/AIDASoft/podio/blob/c7328d62f17061c8360eb2ac0a3fe0ce931fd2ae/include/podio/CollectionIDTable.h#L22-L25

Additionally, for a given category this information will never change, so the `shared_ptr` could be read only once and then kept in the reader itself, rather than building it for every read.

Reply:
Fixed